ir/pattern.h

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2018-present Barefoot Networks, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef IR_PATTERN_H_
+#define IR_PATTERN_H_
+
+#include "ir/ir.h"
+
+/**
+ * Pattern matcher for IR::Expression trees.
+ *
+ * Build pattern expressions with `Pattern` to match subtrees or leaves, and combine
+ * them with C++ operators to build larger patterns that can match tree fragments
+ */
+class Pattern {
+    class Base {
+     public:
+        virtual bool match(const IR::Node *) = 0;
+    } *pattern;
+    Pattern(Base *p) : pattern(p) {}
+    template<class T> class Match : public Base {
+        const T *&m;
```

Comment:
This is very interesting; I was wondering how to do match + binding. This is a nice solution.
We talked about doing this for a long time, but we never got around to doing it...